drivers: counter: Add counter_reset function

rruuaanng opened the pull request:
Add a counter_reset function to the counter subsystem
to reset the counter value to zero. Introduce
a reset interface at the driver layer to implement
the counter_reset function.

Closes: #32157

aescolar commented on include/zephyr/drivers/counter.h:
```diff
@@ -431,6 +433,23 @@ static inline int z_impl_counter_get_value_64(const struct device *dev,
 }
 
+/**
+ * @brief Reset the counter to zero.
```
@nordic-krch would be better to comment here, but what should one expect as driver developer and user for this API? Meaning,
what should one expect to happen with pending alarms after calling this?
Would it be ok to call this while the counter is running or is a user meant to stop if first?

If this only resets the count, maybe it could be called `reset_count`? (reset may be understood to imply to reset more things)

rruuaanng replied:
Based on the user's feedback, what they need is an interface that resets the counter to zero. I named it `counter_reset` to maintain consistency with the prefixes of other APIs, which will help with IDE autocompletion and align better with the user's memory.

rruuaanng replied:
Is it similar to this?
```
 * @brief Reset the counter to the initial value.
```
I think that's good.

rruuaanng commented on include/zephyr/drivers/counter.h:
```diff
@@ -431,6 +433,23 @@ static inline int z_impl_counter_get_value_64(const struct device *dev,
+ * @param dev Pointer to the device structure for the driver instance.
+ *
+ * @retval 0 If successful.
+ * @retval Negative error code on failure getting the counter value.
```
Oh, hah, I copied it from some other section and seem to have forgotten to modify it.

rruuaanng commented on include/zephyr/drivers/counter.h:
```diff
@@ -431,6 +433,23 @@ static inline int z_impl_counter_get_value_64(const struct device *dev,
+{
+	const struct counter_driver_api *api =
+				(struct counter_driver_api *)dev->api;
+
```
I believe this interface should be implemented as a mandatory interface, as resetting the counter is a common operation. WDTY?

nordic-krch replied:
I would add a check. I guess that you won't implement reset for all existing counters. As API existed for few years without this I would not consider it a base/mandatory API.

rruuaanng replied:
Done.

aescolar commented:
@rruuaanng rebase needed

henrikbrixandersen commented on include/zephyr/drivers/counter.h:
```diff
@@ -431,6 +433,27 @@ static inline int z_impl_counter_get_value_64(const struct device *dev,
+				(struct counter_driver_api *)dev->api;
+
+	if (!api->reset) {
+		return -ENOTSUP;
```
As per https://docs.zephyrproject.org/latest/develop/api/design_guidelines.html#return-codes:

```suggestion
		return -ENOSYS;
```

rruuaanng replied:
Yes, I know about it, but other APIs do the same thing. Should I fix them as well? I'm concerned this might cause some unknown side effects. E.g.:
```
static inline int z_impl_counter_get_value_64(const struct device *dev,
					   uint64_t *ticks)
{
	const struct counter_driver_api *api =
				(struct counter_driver_api *)dev->api;

	if (!api->get_value_64) {
		return -ENOTSUP;
	}

	return api->get_value_64(dev, ticks);
}
```

rruuaanng commented on doc/releases/release-notes-4.2.rst:
```diff
@@ -81,6 +81,11 @@ New APIs and options
 
+* Counter
+
+  * Add function :c:func:`counter_reset` to the counter subsystem to reset the counter
+    value to zero. (:github:`32157`)
```
> Counter API allows for counting down so Reset the counter to the initial value?

It seems I didn’t modify its description. A reviewer mentioned it in a previous comment. And I believe `reset` is superior to `clear` because resetting can mean returning to a specific.

rruuaanng commented on include/zephyr/drivers/counter.h:
```diff
@@ -215,6 +215,7 @@ typedef int (*counter_api_get_value)(const struct device *dev,
 				     uint32_t *ticks);
 typedef int (*counter_api_get_value_64)(const struct device *dev,
 			uint64_t *ticks);
+typedef int (*counter_api_reset)(const struct device *dev);
```
In the timer hal enumeration definition, the following information exists:
```
enumerator NRF_TIMER_TASK_CLEAR

    Task for resetting the timer value.
```
It's named `CLEAR`, but its actual meaning is `resetting`. So I believe calling it `reset` better aligns with its original intent.

ref: https://docs.nordicsemi.com/bundle/ncs-2.4.0/page/nrfx/drivers/timer/hal.html

rruuaanng committed:
drivers: counter: Add counter_reset function

Add a counter_reset function to the counter subsystem
to reset the counter value to zero. Introduce
a reset interface at the driver layer to implement
the counter_reset function.

Signed-off-by: James Roy <rruuaanng@outlook.com>

rruuaanng committed:
drivers: counter: Add native-posix implementation example

Implement the '.reset' interface in the native-posix driver
to allow the upper layers to use the counter_reset
function.

Signed-off-by: James Roy <rruuaanng@outlook.com>

rruuaanng committed:
drivers: counter: nrf: Add nrfx-timer implementation

Implement the '.reset' interface in the nrfx-timer driver
to allow the upper layers to use the counter_reset
function.

Signed-off-by: James Roy <rruuaanng@outlook.com>

rruuaanng committed:
doc: releases: Add a counter release entry

Add the counter_reset function in the counter
subsystem to reset the counter value to zero.

Signed-off-by: James Roy <rruuaanng@outlook.com>

rruuaanng committed:
drivers: counter: Fix incorrect error codes

Unimplemented interfaces should return
-ENOSYS instead of -ENOTSUP.

Signed-off-by: James Roy <rruuaanng@outlook.com>
